src/ethereum/starknet/state_root.rs

```diff
@@ -0,0 +1,294 @@
+use std::{collections::VecDeque, time::Duration};
+
+use web3::{transports::WebSocket, types::BlockNumber, Web3};
+
+use crate::ethereum::starknet::{
+    contract::{LogWithOrigin, StateUpdateLog},
+    CoreContract, GetLogsError,
+};
+
+/// Provides a reorg aware stream of L2 state update logs.
+///
+/// ## Important
+///
+/// [StateRootStream] keeps an internal buffer of update logs.
+/// Once these are exhausted, a new set of logs are retrieved from
+/// L1. **Only** at this point can reorgs be detected. This implies
+/// that although this buffer was valid when retrieved, that in the
+/// interim a reorg could have occurred invalidating this buffer.
+///
+/// The implication is that your other L1 / L2 queries will
+/// still need to be reorg aware.
+pub struct StateRootStream {
+    last: Option<LogWithOrigin<StateUpdateLog>>,
+    core_contract: CoreContract,
+    websocket: Web3<WebSocket>,
+    buffer: VecDeque<LogWithOrigin<StateUpdateLog>>,
+    stride: u64,
+}
+
+#[derive(Debug)]
+pub enum GetRootError {
+    /// An L1 chain reorganisation occurred. At the very least
+    /// the last update returned is no longer valid.
+    Reorg,
+    Other(anyhow::Error),
+}
+
+impl From<anyhow::Error> for GetRootError {
+    fn from(err: anyhow::Error) -> Self {
+        GetRootError::Other(err)
+    }
+}
+
+impl From<web3::Error> for GetRootError {
+    fn from(err: web3::Error) -> Self {
+        GetRootError::Other(anyhow::anyhow!("Unexpected error: {}", err))
+    }
+}
+
+impl StateRootStream {
+    /// Creates a new [StateRootStream].
+    ///
+    /// `last_known` should be set to the last known valid [StateUpdateLog],
+    /// or [None] if you are starting from scratch.
+    pub fn new(
+        websocket: Web3<WebSocket>,
+        last_known: Option<LogWithOrigin<StateUpdateLog>>,
+    ) -> Self {
+        let core_contract = CoreContract::load(websocket.clone());
+
+        Self {
+            last: last_known,
+            buffer: VecDeque::new(),
+            core_contract,
+            websocket,
+            stride: 100_000,
+        }
+    }
+
+    /// Returns the next [StateUpdateLog].
+    ///
+    /// If [GetRootError::Reorg] is returned, this [StateRootStream] will no longer
+    /// be valid and should be discarded. One sensible reaction is to validate your
+    /// state and purge whatever has been reorg'd away. Then create a new [StateRootStream]
+    /// starting with your new valid state.
+    pub async fn next(&mut self) -> Result<LogWithOrigin<StateUpdateLog>, GetRootError> {
+        if self.buffer.is_empty() {
+            self.buffer = self.retrieve_logs().await?;
+        }
+        // SAFETY: unwrap is safe due to the above empty check and that
+        // retrieve_next_updates will only return once there is a new update.
+        let last = self.buffer.pop_front().unwrap();
```

Comment:
It's always safe as far as "SAFETY" comments in rust in general go as those are always related to memory safety, panicking and calling `abort(3)` are safe. Would replace this with `.expect("buffer cannot be empty, there should had been an error")`.

Reply:
Was wondering what the appropriate thing was here. Since I would like to document `unwrap`s in general.